src/libpmemfile-posix/inode.c

```diff
@@ -0,0 +1,841 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * inode.c -- inode operations
+ */
+
+#include <errno.h>
+#include <inttypes.h>
+
+#include "callbacks.h"
+#include "data.h"
+#include "dir.h"
+#include "inode.h"
+#include "inode_array.h"
+#include "internal.h"
+#include "locks.h"
+#include "os_thread.h"
+#include "out.h"
+
+/*
+ * pmfi_path -- returns one of the full paths inode can be reached on
+ *
+ * Only for debugging.
+ */
+const char *
+pmfi_path(struct pmemfile_vinode *vinode)
+{
+#ifdef DEBUG
+	if (!vinode)
+		return NULL;
+	if (!vinode->path)
+		LOG(LTRC, "0x%lx: no vinode->path", vinode->tinode.oid.off);
+	return vinode->path;
+#else
+	return NULL;
+#endif
+}
+
+/*
+ * vinode_ref -- increases inode runtime reference counter
+ *
+ * Does not need transaction.
+ */
+struct pmemfile_vinode *
+vinode_ref(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
+{
+	(void) pfp;
+
+	__sync_fetch_and_add(&vinode->ref, 1);
+	return vinode;
+}
+
+#define BUCKET_SIZE 2
+
+/* hash map bucket */
+struct inode_map_bucket {
+	struct {
+		/* persistent inode */
+		TOID(struct pmemfile_inode) pinode;
+
+		/* volatile inode */
+		struct pmemfile_vinode *vinode;
+	} arr[BUCKET_SIZE];
+};
+
+/* inode->vinode hash map */
+struct pmemfile_inode_map {
+	os_rwlock_t rwlock;
+
+	/* hash function coefficients */
+	uint32_t hash_fun_a;
+	uint32_t hash_fun_b;
+	uint64_t hash_fun_p;
+
+	/* number of elements in "buckets" */
+	size_t sz;
+
+	/* buckets */
+	struct inode_map_bucket *buckets;
+
+	/* number of used slots */
+	size_t inodes;
+};
+
+/*
+ * inode_map_rand_params -- randomizes coefficients of the hashing function
+ */
+static void
+inode_map_rand_params(struct pmemfile_inode_map *c)
+{
+	/* XXX use independent random pool */
+	do {
+		c->hash_fun_a = (uint32_t)rand();
+	} while (c->hash_fun_a == 0);
+	c->hash_fun_b = (uint32_t)rand();
+}
+
+/*
+ * inode_map_alloc -- allocates inode hash map
+ */
+struct pmemfile_inode_map *
+inode_map_alloc()
+{
+	struct pmemfile_inode_map *c = calloc(1, sizeof(*c));
+
+	c->sz = 2;
+	c->buckets = calloc(1, c->sz * sizeof(c->buckets[0]));
+
+	inode_map_rand_params(c);
+	c->hash_fun_p = 32212254719ULL;
+
+	os_rwlock_init(&c->rwlock);
+
+	return c;
+}
+
+/*
+ * inode_map_free -- destroys inode hash map
+ */
+void
+inode_map_free(struct pmemfile_inode_map *c)
+{
+	for (unsigned i = 0; i < c->sz; ++i) {
+		struct inode_map_bucket *bucket = &c->buckets[i];
+
+		for (unsigned j = 0; j < BUCKET_SIZE; ++j)
+			if (bucket->arr[j].vinode)
+				FATAL("memory leak");
```

Comment:
This check triggers in 2 scenarios when someone wants to close the pool, but:
* he still have some files open
* library internally messed up reference counting

The first scenario is easy to trigger from tests. The second one is very serious, and that's why for now I want this code to be compiled in for both release and debug builds.